### DISCUSSION TITLE
pb_hooks debugging?

### DISCUSSION THREAD

**Discussion by benallfree** [pocketbase/pocketbase#3342] [Q&A] [answered]
How do we see errors in pb_hooks code? I'm noticing some silent failures with pb_hooks, such as `throw new Error('foo')` doesn't seem to surface anywhere. Also, failing scripts seem not to produce any `stdout` or `stderr` activity from PocketBase.

**Comment by ganigeorgiev** [comment]
Please provide an example (and with which version you are testing).
The error output depends on the hook type (router, cron, middleware handler, etc.).
For most of the cases there should be a stderr log (_you can also start the executable with `--debug` for more detailed output; in the next release the logs will be generalized and you'll be able to view them from the Admin UI, similar to the request logs_).

**Reply by benallfree** [reply]
It's a little hard for me to track down, but:

```ts
// test.pb.js

throw new Error(`foo`)
```

produces

```
Failed to execute test.pb.js:\n - Error: foo at <eval>:4:7(2)
```

But this fails silently on record updates:

```ts
// test.pb.js

onRecordAfterUpdateRequest((e) => {
  throw new Error(`foo`)
})
```

**Reply by ganigeorgiev** [reply]
Throwing an exception or returning an error in a request handler hook is equivalent of returning `error` in Go, aka. it is converted to an API request error (see https://pocketbase.io/docs/js-routing/#error-response).

You'll get a stderr output if you start the executable with `--debug` but as mentioned earlier this will change in the near future and all logs (whenever possible) will be written in the logs db so that they can be explored from the Admin UI.

**Reply by benallfree** [reply]
A little more context... here's the hook that is failing. If I end with `.sql()`, the `console.log` will execute. If I end with `.one(result)`, the final `console.log` never executes and no error information is given.

`onAfterApiError` just says `Something went wrong while processing your request.`

```ts
onRecordAfterUpdateRequest((e) => {
  const lib = require(`${__hooks}/lib.js`) as Lib
  const assert: Assert = lib.assert
  const { record } = e
  assert(record, `Expected record here`)
  const instanceId = record.getString('instanceId')
  console.log(instanceId)

  const now = new Date()

  // Getting the first day of the current month
  const startIso = new Date(now.getFullYear(), now.getMonth(), 1).toISOString()

  // Getting the last day of the current month
  const endIso = new Date(
    now.getFullYear(),
    now.getMonth() + 1,
    0,
  ).toISOString()

  const result = new DynamicModel({
    t: 0,
  })

  const sql = $app
    .dao()
    ?.db()
    .newQuery(
      'SELECT sum(totalSeconds) as t FROM invocations WHERE instanceId={:instanceId} and startedAt>={:startIso} and startedAt<={:endIso}',
    )
    ?.bind({
      instanceId,
      startIso,
      endIso,
    })
    ?.one(result)
  console.log(sql, startIso, endIso, JSON.stringify(result))
}, 'invocations')
```

**Reply by ganigeorgiev** [reply]
Start the executable with `--debug` and/or check the failed request log in the _Admin UI > Logs_.

**Reply by benallfree** [reply]
Ah, ok `--debug` caught something:

```
 sql: Scan error on column index 0, name \"t\": converting driver.Value type float64 (\"537.0100000000001\") to a int64: invalid syntax"
```

So what is the best way to surface these errors for users on PocketHost? It would be nice if they went to `stderr`, otherwise they have to go check their admin log for some errors and stdout for others?

**Reply by ganigeorgiev** [reply]
There is no easy option at the moment without the `--debug` flag.

In v0.19/v0.20 the output of all app related errors (not just request errors) will be written to the `logs.db` and available in the Admin UI.

**Reply by benallfree** [reply]
(clarification) we've gotten requests to capture stdout/stderr, which we write to a separate realtime SQLite db which is exposed through the PocketHost UI so they can see console output. Are you going to capture all PocketBase console output and create this feature in the PocketBase admin?

**Reply by benallfree** [reply]
> In v0.19/v0.20 the output of all app related errors (not just request errors) will be written to the logs.db and available in the Admin UI.

If I could make a suggestion, please consider sending all errors to `stderr` even when not in `--debug` mode. That way users can debug their hooks using `console.log` and `throw new Error` and PocketHost will pick them up on `stderr`.

**Reply by ganigeorgiev** [reply, marked as answer]
> Are you going to capture all PocketBase console output and create this feature in the PocketBase admin?

No, what we are going to do is remove the `app.IsDebug()` checks and instead of calling directly `log.Println(...)` like in https://github.com/pocketbase/pocketbase/blob/f605521208179072a1251042e2b16835fcc955cc/apis/admin.go#L132-L135, we'll have a builtin app logger that will write the result in the database and allow exploration and filtering from the Admin UI.

> please consider sending all errors to stderr even when not in --debug mode.

I'm not sure about this (it kindof conflict with the flag), but even if this is not going to be available out of the box, users will be able to attach to the logs hooks and write to the stdout/stderr (_or anywhere else_) on their own.
